File: frontend/src/services/authService.js
import { getStore, setStore, generateId, delay, KEYS } from './mockData'

export const login = async (email, password) => {
  await delay(300)
  const users = getStore(KEYS.USERS) || []
  const user = users.find((u) => u.email === email.toLowerCase().trim())

  if (!user) throw { response: { status: 401, data: { error: 'Invalid email or password' } } }
  if (user.password !== password) throw { response: { status: 401, data: { error: 'Invalid email or password' } } }
  if (user.is_active === false) throw { response: { status: 403, data: { error: 'Account is deactivated' } } }

  const token = 'mock-jwt-' + generateId()
  const safeUser = { _id: user._id, email: user.email, full_name: user.full_name, role: user.role }

  setStore(KEYS.CURRENT_USER, safeUser)

  return { data: { access_token: token, refresh_token: 'mock-refresh-' + generateId(), user: safeUser, message: 'Login successful' } }
}

export const register = async (email, password, fullName) => {
  await delay(300)
  const users = getStore(KEYS.USERS) || []

  if (users.find((u) => u.email === email.toLowerCase().trim())) {
    throw { response: { status: 409, data: { error: 'Email already registered' } } }
  }

  const newUser = {
    _id: generateId(),
    email: email.toLowerCase().trim(),
    password,
    full_name: fullName,
    role: 'user',
    is_active: true,
    created_at: new Date().toISOString(),
  }

  users.push(newUser)
  setStore(KEYS.USERS, users)

  const token = 'mock-jwt-' + generateId()
  const safeUser = { _id: newUser._id, email: newUser.email, full_name: newUser.full_name, role: newUser.role }

  setStore(KEYS.CURRENT_USER, safeUser)

  return { data: { access_token: token, refresh_token: 'mock-refresh-' + generateId(), user: safeUser, message: 'User registered successfully' } }
}

export const logout = async () => {
  await delay(100)
  localStorage.removeItem(KEYS.CURRENT_USER)
  return { data: { message: 'Successfully logged out' } }
}

export const refreshToken = async () => {
  await delay(100)
  return { data: { access_token: 'mock-jwt-' + generateId(), message: 'Token refreshed' } }
}

File: backend/app/services/auth_service.py
from flask_jwt_extended import create_access_token, create_refresh_token

from app.models.user import UserModel
from app.utils.validators import validate_email, validate_password
from app.utils.helpers import serialize_doc


class AuthService:

    @staticmethod
    def register(email, password, full_name):
        if not validate_email(email):
            return {'error': 'Invalid email format'}, 400

        if not validate_password(password):
            return {'error': 'Password must be at least 8 characters with letters and numbers'}, 400

        if not full_name or len(full_name.strip()) < 2:
            return {'error': 'Full name is required'}, 400

        existing = UserModel.find_by_email(email)
        if existing:
            return {'error': 'Email already registered'}, 409

        user_id = UserModel.create_user(email, password, full_name)
        user = UserModel.find_by_id(user_id)

        access_token = create_access_token(identity=str(user['_id']))
        refresh_token = create_refresh_token(identity=str(user['_id']))

        return {
            'message': 'User registered successfully',
            'access_token': access_token,
            'refresh_token': refresh_token,
            'user': serialize_doc({
                '_id': user['_id'],
                'email': user['email'],
                'full_name': user['full_name'],
                'role': user.get('role', 'user'),
                'created_at': user['created_at'],
            }),
        }, 201

    @staticmethod
    def login(email, password):
        user = UserModel.find_by_email(email)
        if not user:
            return {'error': 'Invalid email or password'}, 401

        if not user.get('is_active', True):
            return {'error': 'Account is deactivated'}, 403

        if not UserModel.verify_password(user['password'], password):
            return {'error': 'Invalid email or password'}, 401

        UserModel.update_last_login(str(user['_id']))

        access_token = create_access_token(identity=str(user['_id']))
        refresh_token = create_refresh_token(identity=str(user['_id']))

        return {
            'message': 'Login successful',
            'access_token': access_token,
            'refresh_token': refresh_token,
            'user': serialize_doc({
                '_id': user['_id'],
                'email': user['email'],
                'full_name': user['full_name'],
                'role': user.get('role', 'user'),
            }),
        }, 200

    @staticmethod
    def refresh_token(identity):
        access_token = create_access_token(identity=identity)
        return {
            'access_token': access_token,
            'message': 'Token refreshed successfully',
        }, 200
